### PR TITLE
Encoding network batch ensemble

### PR DESCRIPTION
**Pull request by runjerry**
Make encoding network work with batch_ensemble layers. 

Handel inputs with ``emsemble_ids`` by wrapping ``input_preprocessors`` and ``preprocessing_combiner`` with ``Parallel``.

### CODE REVIEW

File: alf/networks/encoding_networks.py [hunk 1]
```diff
@@ -700,6 +737,14 @@ def __init__(self,
             assert isinstance(spec, TensorSpec), \
                 "The spec must be an instance of TensorSpec!"
 
+        if input_with_ensemble_ids:
+            nets = [
+                Parallel(
+                    (_Sequential(nets, input_tensor_spec=input_tensor_spec),
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py]: why `_Sequential` instead of `Sequential`?

File: alf/networks/encoding_networks.py [hunk 2]
```diff
@@ -709,11 +754,16 @@ def __init__(self,
             net = ImageEncodingNetwork(
                 input_channels, (height, width),
                 conv_layer_params,
+                use_batch_ensemble=use_batch_ensemble,
+                ensemble_size=ensemble_size,
+                output_ensemble_ids=output_ensemble_ids,
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py]: output_ensemble_ids=True

File: alf/networks/encoding_networks.py [hunk 3]
```diff
@@ -709,11 +754,16 @@ def __init__(self,
...
                 activation=activation,
                 kernel_initializer=kernel_initializer,
                 flatten_output=True)
             spec = net.output_spec
             nets.append(net)
+        if isinstance(spec, tuple):
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py]: `if use_batch_ensemble:`

File: alf/networks/encoding_networks.py [hunk 4]
```diff
@@ -733,9 +783,18 @@ def __init__(self,
             assert isinstance(fc_layer_params, tuple)
             fc_layer_params = list(fc_layer_params)
 
+        if use_batch_ensemble:
+            assert ensemble_size > 1
+            fc_layer_ctor = functools.partial(
+                layers.FCBatchEnsemble,
+                ensemble_size=ensemble_size,
+                output_ensemble_ids=output_ensemble_ids)
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py]: output_ensemble_ids=True

File: alf/networks/encoding_networks.py [hunk 5]
```diff
@@ -755,7 +814,7 @@ def __init__(self,
                 last_kernel_initializer = kernel_initializer
 
             nets.append(
-                layers.FC(
+                fc_layer_ctor(
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py:820]: for batch ensemble: output_ensemble_ids=output_ensemble_ids

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/networks/encoding_networks.py (hunk 2 above)
Comment (runjerry) [reply alf/networks/encoding_networks.py]: Updated.

File: alf/networks/encoding_networks.py (hunk 3 above)
Comment (runjerry) [reply alf/networks/encoding_networks.py]: Updated.

File: alf/networks/encoding_networks.py (hunk 4 above)
Comment (runjerry) [reply alf/networks/encoding_networks.py]: Updated.

File: alf/networks/encoding_networks.py (hunk 5 above)
Comment (runjerry) [reply alf/networks/encoding_networks.py:820]: Updated. Thanks for catching this issue.

File: alf/networks/encoding_networks.py (hunk 1 above)
Comment (runjerry) [reply alf/networks/encoding_networks.py]: The interface of ``Sequential`` wrapper will wrap the input modules into an extra layer of tuple, which fails the ``isinstance(module, Callable)`` check for some of the test cases, say when ``nets = [NestSum()]``.

Comment (emailweixu) [reply alf/networks/encoding_networks.py]: I don't get it. I think `Sequential(*nets, input_tensor_spec=input_tensor_spec)` is equivalent to `_Sequential(nets, input_tensor_spec=input_tensor_spec)`

File: alf/networks/encoding_networks.py (hunk 5 above)
Comment (emailweixu) [reply alf/networks/encoding_networks.py:820]: This means that the unittests do not properly test different scenarios?

Comment (runjerry) [reply alf/networks/encoding_networks.py:820]: That is a good point. I then realized that handling the ``output_ensemble_ids`` option for ``EncodingNetwork`` with and without ``last_layer_size`` becomes a bit tedious, so I just removed this option and always create BatchEnsemble layers with ``output_ensemble_ids=True``.

File: alf/networks/encoding_networks.py (hunk 1 above)
Comment (runjerry) [reply alf/networks/encoding_networks.py]: Oops, forgot the star sign, updated.

File: alf/networks/encoding_networks.py [hunk 6]
```diff
@@ -69,6 +71,9 @@ def __init__(self,
             conv_layer_params (tuppe[tuple]): a non-empty tuple of
                 tuple (num_filters, kernel_size, strides, padding), where
                 padding is optional
+            use_batch_ensemble (bool): whether use BatchEnsemble Conv layers.
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py]: should comment the input and the output in this case.

File: alf/networks/encoding_networks.py [hunk 7]
```diff
@@ -659,6 +682,13 @@ def __init__(self,
                 used.
             use_fc_bn (bool): whether use Batch Normalization for fc layers.
             use_fc_ln (bool): whether use Layer Normalization for fc layers.
+            use_batch_ensemble (bool): whether use BatchEnsemble FC and Conv
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py]: should comment the output of the network, i.e., it is tuple with ensemble_ids

File: alf/networks/encoding_networks.py (hunk 6 above)
Comment (runjerry) [reply alf/networks/encoding_networks.py]: Updated.

File: alf/networks/encoding_networks.py (hunk 7 above)
Comment (runjerry) [reply alf/networks/encoding_networks.py]: Updated.

**Review by emailweixu — APPROVED** [review]

**Pull request merged**